File: geo/text_cleaning.py
"""Lightweight text helpers shared across extraction."""
from __future__ import annotations

import re
from urllib.parse import urlparse

_WORD = re.compile(r"[A-Za-z0-9]")


def _ok_boundary(text: str, start: int, end: int) -> bool:
    """True if the match isn't glued to surrounding alphanumerics/underscore."""
    if start > 0 and (text[start - 1].isalnum() or text[start - 1] == "_"):
        return False
    if end < len(text) and (text[end].isalnum() or text[end] == "_"):
        return False
    return True


def alias_spans(text: str, alias: str) -> list[tuple[int, int]]:
    """Boundary-aware, case-insensitive match spans of `alias` in `text`."""
    if not text or not alias:
        return []
    pattern = re.escape(alias.strip())
    spans = []
    for m in re.finditer(pattern, text, flags=re.IGNORECASE):
        if _ok_boundary(text, m.start(), m.end()):
            spans.append((m.start(), m.end()))
    return spans


def merge_spans(spans: list[tuple[int, int]]) -> list[tuple[int, int]]:
    """Merge overlapping/touching spans so overlapping aliases count once.

    e.g. 'RISA Labs' (0,9) + 'RISA' (0,4) + 'Risa Labs' (0,9) -> one span.
    """
    if not spans:
        return []
    spans = sorted(spans)
    out = [spans[0]]
    for s, e in spans[1:]:
        ls, le = out[-1]
        if s <= le:  # overlap or touch
            out[-1] = (ls, max(le, e))
        else:
            out.append((s, e))
    return out


def count_entity(text: str, aliases: list[str]) -> tuple[int, int | None]:
    """Distinct (de-overlapped) occurrences of an entity across its aliases.

    Returns (count, first_start_index_or_None).
    """
    spans: list[tuple[int, int]] = []
    for a in aliases:
        spans.extend(alias_spans(text, a))
    merged = merge_spans(spans)
    return len(merged), (merged[0][0] if merged else None)


def domain_of(url: str) -> str:
    try:
        net = urlparse(url if "://" in url else "http://" + url).netloc.lower()
        return net[4:] if net.startswith("www.") else net
    except Exception:
        return ""


def domain_root(domain: str) -> str:
    """e.g. 'coherehealth.com' -> 'coherehealth' (the registrable label-ish)."""
    parts = domain.split(".")
    return parts[-2] if len(parts) >= 2 else (parts[0] if parts else "")

File: web/lib/pipeline/textCleaning.ts
// Lightweight text helpers shared across extraction. Ported from geo/text_cleaning.py.

function okBoundary(text: string, start: number, end: number): boolean {
  const isWord = (ch: string) => /[A-Za-z0-9_]/.test(ch);
  if (start > 0 && isWord(text[start - 1])) return false;
  if (end < text.length && isWord(text[end])) return false;
  return true;
}

function escapeRe(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function aliasSpans(text: string, alias: string): [number, number][] {
  if (!text || !alias) return [];
  const spans: [number, number][] = [];
  const re = new RegExp(escapeRe(alias.trim()), "gi");
  let m: RegExpExecArray | null;
  while ((m = re.exec(text)) !== null) {
    const start = m.index;
    const end = m.index + m[0].length;
    if (okBoundary(text, start, end)) spans.push([start, end]);
    if (m.index === re.lastIndex) re.lastIndex++; // guard against zero-width
  }
  return spans;
}

export function mergeSpans(spans: [number, number][]): [number, number][] {
  if (!spans.length) return [];
  const sorted = [...spans].sort((a, b) => a[0] - b[0] || a[1] - b[1]);
  const out: [number, number][] = [sorted[0]];
  for (const [s, e] of sorted.slice(1)) {
    const last = out[out.length - 1];
    if (s <= last[1]) last[1] = Math.max(last[1], e);
    else out.push([s, e]);
  }
  return out;
}

// Distinct (de-overlapped) occurrences of an entity across its aliases.
export function countEntity(text: string, aliases: string[]): [number, number | null] {
  const spans: [number, number][] = [];
  for (const a of aliases) spans.push(...aliasSpans(text, a));
  const merged = mergeSpans(spans);
  return [merged.length, merged.length ? merged[0][0] : null];
}

export function domainOf(url: string): string {
  try {
    const u = new URL(url.includes("://") ? url : "http://" + url);
    const net = u.hostname.toLowerCase();
    return net.startsWith("www.") ? net.slice(4) : net;
  } catch {
    return "";
  }
}

export function domainRoot(domain: string): string {
  const parts = domain.split(".");
  return parts.length >= 2 ? parts[parts.length - 2] : parts[0] || "";
}
